nixos/modules/services/networking/godns.nix

```diff
@@ -0,0 +1,87 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  inherit (lib)
+    mkEnableOption
+    mkIf
+    mkOption
+    mkPackageOption
+    types
+    ;
+
+  cfg = config.services.godns;
+
+  configFormat = pkgs.formats.yaml { };
+  configFile =
+    if cfg.configFile == null then configFormat.generate "config.yaml" cfg.config else cfg.configFile;
+in
+{
+
+  options.services.godns = {
+    enable = mkEnableOption "GoDNS service";
+
+    package = mkPackageOption pkgs "godns" { };
+
+    config = mkOption {
+      type = lib.types.submodule {
+        freeformType = configFormat.type;
+      };
+
+      default = { };
+
+      description = ''
+        Configuration for GoDNS. Refer to the [configuration section](1) in the
+        GoDNS GitHub repository for details.
+
+        [1]: https://github.com/TimothyYe/godns?tab=readme-ov-file#configuration
+      '';
+    };
+
+    configFile = lib.mkOption {
+      default = null;
+      description = ''
+        Path to a custom GoDNS configuration file.
+
+        If set, this option overrides the configuration provided by the `config`
+        option. This is particularly useful for specifying sensitive values like
+        the `login_token` for API providers. For generating a configuration file
+        with secrets, consider using [sops-nix templates][1].
+
+        [1]: https://github.com/Mic92/sops-nix?tab=readme-ov-file#templates
+      '';
+      type = types.nullOr types.path;
+    };
+
+    additionalRestartTriggers = mkOption {
+      default = [ ];
+      type = types.listOf types.unspecified;
+      description = ''
+        Additional triggers to restart the GoDNS service.
+
+        This can be used to restart the service when, for example, a secret used
+        to generate the configuration changes.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.godns = {
+      description = "GoDNS service";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      serviceConfig = {
+        ExecStart = "${lib.getExe cfg.package} -c ${configFile}";
+        Restart = "always";
+        KillMode = "process";
```

Comment:
What's wrong with the default `KillMode`?
> Note that it is not recommended to set KillMode= to process or even none, as this allows processes to escape the service manager's lifecycle and resource management, and to remain running even while their service is considered stopped and is assumed to not consume any resources.

https://www.freedesktop.org/software/systemd/man/latest/systemd.kill.html#KillMode=